security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,49 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	"hash"
+)
+
+type S2AHalfConnection struct {
+	aeadCrypter s2aAeadCrypter
+	expander    hkdfExpander
+	sequence    counter
+}
+
+func NewHalfConn(aeadCrypter s2aAeadCrypter, expander hkdfExpander) S2AHalfConnection {
+	return S2AHalfConnection{aeadCrypter, expander, newCounter()}
+}
+
+func (hc *S2AHalfConnection) Encrypt(dst, plaintext, aad []byte) ([]byte, error) {
+	// TODO(rnkim): Implement this.
+	panic("Encrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) Decrypt(dst, ciphertext, aad []byte) ([]byte, error) {
+	// TODO(rnkim): Implement this.
+	panic("Decrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) UpdateKey(key []byte) error {
+	// TODO(rnkim): Implement this.
+	panic("UpdateKey currently unimplemented")
+}
+
+// expandLabel implements HKDF-Expand-Label specified in
+// https://tools.ietf.org/html/rfc8446#section-7.1. Note that the `Context` and
+// `Length` parameters have been omitted since we always pass in an empty string
+// for `Context` and we used a fixed length.
+func (hc *S2AHalfConnection) expandLabel(h func() hash.Hash, secret, label []byte) ([]byte, error) {
```

Comment:
I think we can model this function off the [C++ derive_secret API](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L119). To do so, we would not add "tls13 " to `label`. Rather, the caller will plug in values like "tls13 key" or "tls13 upd" for `label` when the caller wants to derive a key, or do a key update.

Reply:
Was there any specific reason the C++ version was modeled this way? I believe "tls13 " is always a prefix for the HkdfLabel specified in the TLS 1.3 RFC. It seems unnecessary to prepend "tls13 " in front of every label when it will never change.

Reply:
For C++, the reason was just that it was simpler - in S2A-Go terminology, the choices were:

1. There is  `expandLabel` (implemented using HKDF-expand), called by `deriveSecret`, which is called by `deriveKey`, `deriveNonce`, and `advanceTrafficSecret`.
2. There is `deriveSecret` (implemented using HKDF-expand), which is called by `deriveKey`, `deriveNonce`, and `advanceTrafficSecret`.

Option 2 was used because the `expandLabel` API wasn't needed: we had to put in a different label for each of the 3 uses, so might as well put in the label prefix as well.

Reply:
Thanks for the explanation! Do you think it would be worth passing in the entire label with the "tl13 " prepended to it like in C++? I think it might be worth saving a couple of bytes here by not passing it to every call of `expandLabel` or `deriveSecret`, but there's also some value in keeping it consistent with the C++ version as well.

Reply:
I think there's probably no real performance difference, unless adding a call to the stack trace is more expensive in Go than I've been led to believe. :)

I leave it up to you. Personally, I think removing the `expandLabel` and only having `deriveSecret` (that takes in the full label) is easier for a reader to understand. But I don't have very strong feelings either way.